src/Twig/Extension/SonataAdminExtension.php

```diff
@@ -521,293 +553,231 @@ public function getUrlSafeIdentifier($model, ?AdminInterface $admin = null)
-            $object = new FieldVote($object, $field);
-        }
-
-        if (!\is_array($role)) {
-            $role = [$role];
+        if ('sonata_deprecation_mute' !== (\func_get_args()[3] ?? null)) {
+            @trigger_error(sprintf(
+                'The %s method is deprecated in favor of SecurityExtension::isGrantedAffirmative since version 3.x and will be removed in 4.0.',
+                __METHOD__
+            ), E_USER_DEPRECATED);
         }
 
-        foreach ($role as $oneRole) {
-            try {
-                if ($this->securityChecker->isGranted($oneRole, $object)) {
-                    return true;
-                }
-            } catch (AuthenticationCredentialsNotFoundException $e) {
-                // empty on purpose
-            }
+        if (null === $this->securityExtension) {
+            $this->securityExtension = new SecurityExtension($this->securityChecker);
         }
 
-        return false;
+        return $this->securityExtension->isGrantedAffirmative($role, $object, $field);
     }
 
     /**
+     * NEXT_MAJOR: Remove this method.
+     *
      * Get template.
      *
      * @param string $defaultTemplate
      *
      * @return TemplateWrapper
+     *
+     * @deprecated since sonata-project/admin-bundle 3.x and will be removed in 4.0
      */
     protected function getTemplate(
         FieldDescriptionInterface $fieldDescription,
         $defaultTemplate,
         Environment $environment
     ) {
-        $templateName = $fieldDescription->getTemplate() ?: $defaultTemplate;
-
-        try {
-            $template = $environment->load($templateName);
-        } catch (LoaderError $e) {
+        if ('sonata_deprecation_mute' !== (\func_get_args()[3] ?? null)) {
             @trigger_error(sprintf(
-                'Relying on default template loading on field template loading exception is deprecated since 3.1'
-                .' and will be removed in 4.0. A %s exception will be thrown instead',
-                LoaderError::class
+                'The %s method is deprecated in favor of RenderElementExtension::getTemplate since version 3.x and will be removed in 4.0.',
+                __METHOD__
             ), E_USER_DEPRECATED);
-            $template = $environment->load($defaultTemplate);
-
-            if (null !== $this->logger) {
-                $this->logger->warning(sprintf(
-                    'An error occured trying to load the template "%s" for the field "%s",'
-                    .' the default template "%s" was used instead.',
-                    $templateName,
-                    $fieldDescription->getFieldName(),
-                    $defaultTemplate
-                ), ['exception' => $e]);
-            }
         }
 
-        return $template;
+        return $this->renderElementExtension->getTemplate($fieldDescription, $defaultTemplate, $$environment);
     }
 
+    /**
+     * NEXT_MAJOR: Remove this method.
+     */
     private function render(
```

Comment:
IMO I would remove this method from here since it's private and I would make it also private in `RenderElementExtension`. If I understood correctly, this is for the deprecated `SonataAdminExtension::output()` method. Since it's deprecated, I would copy the content of `render` to that method since it's small and is going to be deleted.